cmd/upgrade/app/v1alpha1/result.go

```diff
@@ -0,0 +1,271 @@
+	urb.Tasks = tasks
+	return urb
+}
+
+// validate validates UpgradeResultGetOrCreateBuilder instance
+func (urb *UpgradeResultGetOrCreateBuilder) validate() error {
+	if len(urb.errors) != 0 {
+		return errors.Errorf(
+			"failed to validate: build errors were found: %v", urb.errors)
+	}
+	validationErrs := []error{}
+	if urb.SelfName == "" {
+		validationErrs = append(validationErrs,
+			errors.New("missing self name"))
+	}
+	if urb.SelfNamespace == "" {
+		validationErrs = append(validationErrs,
+			errors.New("missing self namespace"))
+	}
+	if urb.SelfUID == "" {
+		validationErrs = append(validationErrs,
+			errors.New("missing self uid"))
+	}
+	if urb.UpgradeConfig == nil {
+		validationErrs = append(validationErrs,
+			errors.New("missing upgrade config"))
+	}
+	if urb.ResourceDetails == nil {
+		validationErrs = append(validationErrs,
+			errors.New("missing resource details"))
+	}
+	if len(urb.Tasks) == 0 {
+		validationErrs = append(validationErrs,
+			errors.New("missing tasks"))
+	}
+	if len(validationErrs) != 0 {
+		urb.errors = append(urb.errors, validationErrs...)
+		return errors.Errorf("validation error(s) found: %v", validationErrs)
+	}
+	return nil
+}
+
+// GetOrCreate builds a new instance of UpgradeResult with the
+// helps of UpgradeResultGetOrCreateBuilder. Upgrade result cr
+// is required to maintain resiliency in upgrade.
+func (urb *UpgradeResultGetOrCreateBuilder) GetOrCreate() (
+	res *upgrade.UpgradeResult, err error) {
+	err = urb.validate()
+	if err != nil {
+		return nil,
+			errors.Wrapf(err, "failed to build UpgradeResultCR: %s", urb)
+	}
+	l := labelJobName + "=" + urb.SelfName +
+		"," + labelItemName + "=" + urb.ResourceDetails.Name +
+		"," + labelItemNamespace + "=" + urb.ResourceDetails.Namespace +
+		"," + labelItemKind + "=" + urb.ResourceDetails.Kind
+	opts := metav1.ListOptions{
+		LabelSelector: l,
+	}
+	urList, err := upgraderesult.KubeClient(
+		upgraderesult.WithNamespace(urb.SelfNamespace)).
+		List(opts)
+	if err != nil {
+		return nil,
+			errors.Wrapf(err, "failed to build UpgradeResultCR: %s", urb)
+	}
+	if len(urList.Items) == 0 {
+		ur, err := urb.getUpgradeResultObj()
+		if err != nil {
+			return nil,
+				errors.Wrapf(err, "failed to build UpgradeResultCR: %s", urb)
+		}
+
+		urb.UpgradeResult.object, err = upgraderesult.KubeClient(
+			upgraderesult.WithNamespace(urb.SelfNamespace)).
+			Create(ur)
+		if err != nil {
+			return nil,
+				errors.Wrapf(err, "failed to build UpgradeResultCR: %s", urb)
+		}
+		return urb.UpgradeResult.object, nil
+	} else if len(urList.Items) == 1 {
+		return &urList.Items[0], nil
+	}
+	return nil,
+		errors.Errorf(`failed to build UpgradeResultCR:
+		multiple upgrade result cr found for resource: %v
+		upgrade result crs: %v`, urb.ResourceDetails, urList)
+}
+
+// getUpgradeResultObj returns UpgradeResult Object for given resource
+func (urb *UpgradeResultGetOrCreateBuilder) getUpgradeResultObj() (
```

Comment:
Can we rename this to `buildUpgradeResult`

Reply:
done